File: packages/icoder-python/icoder_sdk/resources/documents.py
"""Corti-compatible Classic Documents generation and lifecycle."""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

from ..client import iCoDerClient
from ..request_options import RequestOptions


class DocumentsResource:
    def __init__(self, client: iCoDerClient):
        self._client = client

    @staticmethod
    def _base(interaction_id: str) -> str:
        return f"/api/v2/tools/interactions/{quote(interaction_id, safe='')}/documents"

    def create(
        self,
        interaction_id: str,
        request: dict[str, Any],
        *,
        retention_policy: str | None = None,
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        if retention_policy not in {None, "none"}:
            raise ValueError("retention_policy supports only None or 'none'")
        headers = (
            {"X-Corti-Retention-Policy": "none"}
            if retention_policy == "none"
            else None
        )
        response = self._client.post(
            f"{self._base(interaction_id)}/",
            json=request,
            headers=headers,
            request_options=request_options,
        )
        response.raise_for_status()
        return {
            "document": response.json(),
            "status_code": response.status_code,
            "retention_acknowledged": (
                response.headers.get("X-Corti-Retention-Policy") == "acknowledged"
            ),
        }

    def preview(
        self,
        interaction_id: str,
        request: dict[str, Any],
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        result = self.create(
            interaction_id,
            request,
            retention_policy="none",
            request_options=request_options,
        )
        if not result["retention_acknowledged"]:
            raise RuntimeError("server did not acknowledge the zero-retention policy")
        return result["document"]

    def list(
        self, interaction_id: str, request_options: RequestOptions | None = None,
    ) -> list[dict[str, Any]]:
        response = self._client.get(
            f"{self._base(interaction_id)}/", request_options=request_options,
        )
        response.raise_for_status()
        return response.json()["data"]

    def get(
        self,
        interaction_id: str,
        document_id: str,
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        response = self._client.get(
            f"{self._base(interaction_id)}/{quote(document_id, safe='')}",
            request_options=request_options,
        )
        response.raise_for_status()
        return response.json()

    def update(
        self,
        interaction_id: str,
        document_id: str,
        request: dict[str, Any],
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        response = self._client.patch(
            f"{self._base(interaction_id)}/{quote(document_id, safe='')}",
            json=request,
            request_options=request_options,
        )
        response.raise_for_status()
        return response.json()

    def delete(
        self,
        interaction_id: str,
        document_id: str,
        request_options: RequestOptions | None = None,
    ) -> None:
        response = self._client.delete(
            f"{self._base(interaction_id)}/{quote(document_id, safe='')}",
            request_options=request_options,
        )
        response.raise_for_status()

File: packages/icoder-sdk/src/resources/documents.ts
import type { AxiosInstance } from 'axios';
import { requestConfig, type iCoDerRequestOptions } from '../request-options.js';

export interface DocumentFact {
  text: string;
  group?: string;
  source?: 'core' | 'system' | 'user';
}

export type DocumentContext =
  | { type: 'facts'; data: DocumentFact[] }
  | { type: 'transcript'; data: { text: string; channel?: number; participant?: number; speakerId?: number; start?: number; end?: number } }
  | { type: 'string'; data: string };

export interface DocumentSectionOverride {
  key: string;
  nameOverride?: string;
  writingStyleOverride?: string;
  formatRuleOverride?: string;
  additionalInstructionsOverride?: string;
  contentOverride?: string;
}

export type DocumentTemplate =
  | {
      sections: DocumentSectionOverride[];
      sectionKeys?: never;
      description?: string;
      documentName?: string;
      additionalInstructionsOverride?: string;
    }
  | {
      sectionKeys: string[];
      sections?: never;
      documentName?: string;
      additionalInstructions?: string;
    };

interface DocumentCreateBase {
  context: DocumentContext[];
  name?: string;
  outputLanguage: string;
  disableGuardrails?: boolean;
  documentationMode?: 'global_sequential' | 'routed_parallel';
}

export type DocumentCreateRequest = DocumentCreateBase & (
  | { templateKey: string; template?: never }
  | { template: DocumentTemplate; templateKey?: never }
);

export interface DocumentSection {
  key: string;
  name: string;
  text: string;
  sort: number;
  createdAt: string;
  updatedAt: string;
}

export interface ClassicDocument {
  id: string;
  name: string;
  templateRef: string;
  isStream: boolean;
  sections: DocumentSection[];
  createdAt: string;
  updatedAt: string;
  outputLanguage: string;
  usageInfo: { creditsConsumed: number };
}

export interface DocumentCreateResult {
  document: ClassicDocument;
  statusCode: number;
  retentionAcknowledged: boolean;
}

export class DocumentsResource {
  constructor(private readonly http: AxiosInstance) {}

  async create(
    interactionId: string,
    request: DocumentCreateRequest,
    options: { retentionPolicy?: 'none' } = {},
    requestOptions?: iCoDerRequestOptions,
  ): Promise<DocumentCreateResult> {
    const response = await this.http.post<ClassicDocument>(
      `/api/v2/tools/interactions/${encodeURIComponent(interactionId)}/documents/`,
      request,
      requestConfig(
        requestOptions,
        {},
        options.retentionPolicy === 'none'
          ? { 'X-Corti-Retention-Policy': 'none' }
          : {},
      ),
    );
    return {
      document: response.data,
      statusCode: response.status,
      retentionAcknowledged:
        response.headers['x-corti-retention-policy'] === 'acknowledged',
    };
  }

  async preview(
    interactionId: string,
    request: DocumentCreateRequest,
    requestOptions?: iCoDerRequestOptions,
  ): Promise<ClassicDocument> {
    const result = await this.create(
      interactionId, request, { retentionPolicy: 'none' }, requestOptions,
    );
    if (!result.retentionAcknowledged) {
      throw new Error('Server did not acknowledge the zero-retention policy');
    }
    return result.document;
  }

  async list(interactionId: string, options?: iCoDerRequestOptions): Promise<ClassicDocument[]> {
    const { data } = await this.http.get<{ data: ClassicDocument[] }>(
      `/api/v2/tools/interactions/${encodeURIComponent(interactionId)}/documents/`,
      requestConfig(options),
    );
    return data.data;
  }

  async get(
    interactionId: string,
    documentId: string,
    options?: iCoDerRequestOptions,
  ): Promise<ClassicDocument> {
    const { data } = await this.http.get<ClassicDocument>(
      `/api/v2/tools/interactions/${encodeURIComponent(interactionId)}/documents/${encodeURIComponent(documentId)}`,
      requestConfig(options),
    );
    return data;
  }

  async update(
    interactionId: string,
    documentId: string,
    request: { name?: string; sections?: Array<Pick<DocumentSection, 'key' | 'name' | 'text' | 'sort'>> },
    options?: iCoDerRequestOptions,
  ): Promise<ClassicDocument> {
    const { data } = await this.http.patch<ClassicDocument>(
      `/api/v2/tools/interactions/${encodeURIComponent(interactionId)}/documents/${encodeURIComponent(documentId)}`,
      request,
      requestConfig(options),
    );
    return data;
  }

  async delete(
    interactionId: string,
    documentId: string,
    options?: iCoDerRequestOptions,
  ): Promise<void> {
    await this.http.delete(
      `/api/v2/tools/interactions/${encodeURIComponent(interactionId)}/documents/${encodeURIComponent(documentId)}`,
      requestConfig(options),
    );
  }
}
